### PR TITLE
Add labels for all prometheus metrics

### PR DESCRIPTION
**Pull request by cydrain**

### CODE REVIEW

File: include/knowhere/prometheus_client.h [hunk 1]
```diff
@@ -55,22 +55,35 @@ class PrometheusClient {
 extern const prometheus::Histogram::BucketBoundaries defaultBuckets;
 extern const std::unique_ptr<PrometheusClient> prometheusClient;
 
-#define DEFINE_PROMETHEUS_GAUGE(name, desc)                                                                  \
-    prometheus::Family<prometheus::Gauge>& name##_family =                                                   \
-        prometheus::BuildGauge().Name(#name).Help(desc).Register(knowhere::prometheusClient->GetRegistry()); \
+#define KNOWHERE_LABEL "knowhere"
```
Comment (zhengbuqian) [inline include/knowhere/prometheus_client.h:58]: PROMETHEUS_KNOWHERE_LABEL

File: include/knowhere/prometheus_client.h [hunk 2]
```diff
@@ -55,22 +55,35 @@ class PrometheusClient {
...
+
+#define DEFINE_PROMETHEUS_GAUGE(name, desc, label)                                                                   \
+    prometheus::Family<prometheus::Gauge>& name##_family = prometheus::BuildGauge()                                  \
```
Comment (zhengbuqian) [inline include/knowhere/prometheus_client.h:61]: it seems to me a bad idea to create a new family for every new metric? for example can we have a latency family and create multiple gauges in this family for search/build/load latency respectively.

File: include/knowhere/prometheus_client.h [hunk 3]
```diff
@@ -55,22 +55,35 @@ class PrometheusClient {
 extern const prometheus::Histogram::BucketBoundaries defaultBuckets;
 extern const std::unique_ptr<PrometheusClient> prometheusClient;
 
-#define DEFINE_PROMETHEUS_GAUGE(name, desc)                                                                  \
-    prometheus::Family<prometheus::Gauge>& name##_family =                                                   \
-        prometheus::BuildGauge().Name(#name).Help(desc).Register(knowhere::prometheusClient->GetRegistry()); \
+#define KNOWHERE_LABEL "knowhere"
+
+#define DEFINE_PROMETHEUS_GAUGE(name, desc, label)                                                                   \
```
Comment (zhengbuqian) [inline include/knowhere/prometheus_client.h:60]: what could be other possible values for label? cardinal? I wonder can we avoid repeating KNOWHERE_LABEL in every usage of DEFINE_PROMETHEUS_GAUGE.

Comment (cydrain) [reply include/knowhere/prometheus_client.h:60]: We can also use these MACROs in Milvus segcore, label will be filled with other values there

File: include/knowhere/prometheus_client.h (hunk 1 above)
Comment (cydrain) [reply include/knowhere/prometheus_client.h:58]: OK

File: include/knowhere/prometheus_client.h (hunk 3 above)
Comment (zhengbuqian) [inline include/knowhere/prometheus_client.h:60]: use another name than label(maybel "module"?), label refers to a prometheus label

File: include/knowhere/prometheus_client.h [hunk 4]
```diff
@@ -55,38 +55,53 @@ class PrometheusClient {
...
+#define DEFINE_PROMETHEUS_HISTOGRAM_WITH_BUCKETS(name, module, buckets) \
+    prometheus::Histogram& CONCATENATE(module, name) = CONCATENATE(name, family).Add({{"module", #module}}, buckets);
+
+#define DEFINE_PROMETHEUS_HISTOGRAM(name, module) DEFINE_PROMETHEUS_HISTOGRAM_WITH_BUCKETS(name, module, defaultBuckets)
+
+#define DECLARE_PROMETHEUS_GAUGE(name, module) extern prometheus::Gauge& CONCATENATE(module, name);
+#define DECLARE_PROMETHEUS_COUNTER(name, module) extern prometheus::Counter& CONCATENATE(module, name);
+#define DECLARE_PROMETHEUS_HISTOGRAM(name, module) extern prometheus::Histogram& CONCATENATE(module, name);
+
+DECLARE_PROMETHEUS_HISTOGRAM(build_latency, PROMETHEUS_LABEL_KNOWHERE);
```
Comment (foxspy) [inline include/knowhere/prometheus_client.h:89]: pls help add cardinal label for build_latency and load_latency

File: include/knowhere/prometheus_client.h [hunk 5]
```diff
@@ -55,38 +55,53 @@ class PrometheusClient {
...
+DECLARE_PROMETHEUS_HISTOGRAM(search_latency, PROMETHEUS_LABEL_CARDINAL);
+
+DECLARE_PROMETHEUS_HISTOGRAM(range_search_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(range_search_latency, PROMETHEUS_LABEL_CARDINAL);
+
+DECLARE_PROMETHEUS_HISTOGRAM(ann_iterator_init_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(search_topk, PROMETHEUS_LABEL_KNOWHERE);
+
+DECLARE_PROMETHEUS_HISTOGRAM(hnsw_bitset_ratio, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(hnsw_search_hops, PROMETHEUS_LABEL_KNOWHERE);
```
Comment (foxspy) [inline include/knowhere/prometheus_client.h:102]: pls help add a cardinal_bitset_ratio metric with cardinal label

File: include/knowhere/prometheus_client.h [hunk 6]
```diff
@@ -55,38 +55,53 @@ class PrometheusClient {
...
+DECLARE_PROMETHEUS_HISTOGRAM(build_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(load_latency, PROMETHEUS_LABEL_KNOWHERE);
+
+DECLARE_PROMETHEUS_HISTOGRAM(search_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(search_latency, PROMETHEUS_LABEL_CARDINAL);
+
+DECLARE_PROMETHEUS_HISTOGRAM(range_search_latency, PROMETHEUS_LABEL_KNOWHERE);
+DECLARE_PROMETHEUS_HISTOGRAM(range_search_latency, PROMETHEUS_LABEL_CARDINAL);
+
+DECLARE_PROMETHEUS_HISTOGRAM(ann_iterator_init_latency, PROMETHEUS_LABEL_KNOWHERE);
```
Comment (foxspy) [inline include/knowhere/prometheus_client.h:98]: pls help add cardinal label for these ann_iterator_init_latency and search_topk metric

File: include/knowhere/prometheus_client.h (hunk 4 above)
Comment (cydrain) [reply include/knowhere/prometheus_client.h:89]: done

File: include/knowhere/prometheus_client.h (hunk 6 above)
Comment (cydrain) [reply include/knowhere/prometheus_client.h:98]: done

**Review by foxspy — APPROVED** [review]
/lgtm

**Pull request merged**